vncviewer/win32TouchHandler.cxx

```diff
@@ -0,0 +1,369 @@
+    dwFlags = MOUSEEVENTF_WHEEL;
+    mouseData = WHEEL_DELTA / 2;
+    break;
+  case 5:
+    dwFlags = MOUSEEVENTF_WHEEL;
+    mouseData = -WHEEL_DELTA / 2;
+    break;
+  case 6:
+    dwFlags = 0x01000; //MOUSEEVENTF_HWHEEL
+    mouseData = WHEEL_DELTA / 2;
+    break;
+  case 7:
+    dwFlags = 0x01000; //MOUSEEVENTF_HWHEEL
+    mouseData = -WHEEL_DELTA / 2;
+    break;
+  }
+
+  pushFakeEvent(type, MOUSEMOVE_FLAGS, mouseData, pos);
+  pushFakeEvent(type, dwFlags, mouseData, pos);
+}
+
+void win32TouchHandler::fakeKeyEvent(bool press, int keysym,
+                                     const GHEvent origEvent) {
+  POINT pos[] = { (long)origEvent.event_x, (long)origEvent.event_y };
+  vlog.error("queuing key%s: 0x%x", press ? "down" : "up", VK_CONTROL);
+  pushFakeEvent(INPUT_KEYBOARD, press ? 0 : KEYEVENTF_KEYUP, VK_CONTROL, pos);
+}
+
+void win32TouchHandler::pushFakeEvent(DWORD type, DWORD dwFlags, DWORD data,
+                                      POINT pos[]) {
+  INPUT input;
+  POINT absPos[] = { pos->x, pos->y };
+
+  // Convert to absolute coordinates for mouse wheel events
+  // FIXME: THIS MUST BE TESTED WITH DUAL SCREEN
+  ClientToScreen(hWnd, absPos);
+  // The absolute screen coordinates need to be converted to
+  // a number between 0 and 65535 for some reason.
+  absPos->x = absPos->x * (65536 / GetSystemMetrics(SM_CXSCREEN));
+  absPos->y = absPos->y * (65536 / GetSystemMetrics(SM_CYSCREEN));
+
+  input.type = type;
+  if (type == INPUT_KEYBOARD) {
+    vlog.error("queuing keyboard 0x%x", (unsigned int)data);
+    input.ki.wScan = 0;
+    input.ki.time = 0;
+    input.ki.dwExtraInfo = 0;
+    input.ki.wVk = data;
+    input.ki.dwFlags = dwFlags;
+  } else if (type == INPUT_MOUSE) {
+    vlog.error("queuing mouse 0x%x (%i, %i)", (unsigned int)dwFlags, (int)absPos->x, (int)absPos->y);
+    input.mi.mouseData = data;
+    input.mi.time = 0;
+    input.mi.dx = absPos->x;
+    input.mi.dy = absPos->y;
+    input.mi.dwFlags = dwFlags;
+  }
+  inputQueue.push_back(input);
+  if (!inputTimer.isStarted())
+    inputTimer.start(INPUT_TIMEOUT);
+}
+
+bool win32TouchHandler::handleTimeout(rfb::Timer* t)
+{
+  if (t == &inputTimer)
+    pushQueuedFakeEvents();
+
+  return false;
+}
+
+void win32TouchHandler::pushQueuedFakeEvents() {
+  int inputSize = inputQueue.size();
+  INPUT inputs[inputSize];
+
+  if (inputQueue.size() <= 0) {
+    return;
+  }
+  for (int i = 0; i < inputSize; i++) {
+    inputs[i] = inputQueue.front();
+    if (inputs[i].type == INPUT_MOUSE) {
+      vlog.error("sending mouse event");
+    } else if (inputs[i].type == INPUT_KEYBOARD) {
+      vlog.error("sending keyboard event");
+    }
+    // FIXME
+    //SendInput(1, &inputs[i], sizeof(INPUT));
+    //Sleep(1000); //messages are processed AFTER the sleep if within the same gesture??
+    inputQueue.pop_front();
+  }
+
+  if (SendInput(inputSize, inputs, sizeof(INPUT)) == 0) {
+    vlog.error("Couldn't send input: 0x%x", (unsigned int)GetLastError());
+  }
```

Comment:
Changing to SendInput() instead of PostMessage doesn't help w.r.t the order of mouse/keyboard events. Even if SendInput() sends a list of inputs in one call:

 MouseMove
 Ctrl KeyPress
 MouseScroll ButtonPress
 MouseScroll ButtonRelease
 Ctrl KeyRelease

The result inside the session will be:

 MouseMove
 MouseScroll ButtonPress
 MouseScroll ButtonRelease
 Ctrl KeyRelease
 Ctrl KeyPress

Reply:
The problem was our AltGr handling on windows. Fixed now.

Reply:
I dropped the SendInput commit.